kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/AbstractProducerMetrics.java

```diff
@@ -0,0 +1,262 @@
+  protected static final String REQUEST_SENSOR_NAME = "request-sensor";
+  protected static final String REQUEST_SIZE_SENSOR_NAME = "request-size-sensor";
+  protected static final String RESPONSE_SENSOR_NAME = "response-sensor";
+  protected static final String RECORD_ERROR_SENSOR_NAME = "record-error-sensor";
+  protected static final String REQUEST_LATENCY_SENSOR_NAME = "request-latency-sensor";
+
+  // request
+  static final String REQUEST_RATE_METRIC_NAME = "request-rate";
+  private static final String REQUEST_RATE_METRIC_DOC =
+      "The average number of requests sent per second.";
+
+  static final String REQUEST_SIZE_AVG_METRIC_NAME = "request-size-avg";
+  private static final String REQUEST_SIZE_AVG_METRIC_DOC = "The average request size in bytes.";
+
+  protected static final String REQUEST_SIZE_WINDOWED_SUM_METRIC_NAME = "request-size-windowed-sum";
+  private static final String REQUEST_SIZE_WINDOWED_SUM_METRIC_DOC =
+      "The summed size in bytes of requests sent within the given window";
+
+  static final String REQUEST_COUNT_WINDOWED_METRIC_NAME = "request-count-windowed";
+  private static final String REQUEST_COUNT_WINDOWED_METRIC_DOC =
+      "The total number of requests sent within the given window.";
+
+  // response
+  static final String RESPONSE_RATE_METRIC_NAME = "response-rate";
+  private static final String RESPONSE_RATE_METRIC_DOC =
+      "The average number of responses sent per second.";
+
+  static final String RESPONSE_COUNT_WINDOWED_METRIC_NAME = "response-count-windowed";
+  private static final String RESPONSE_COUNT_WINDOWED_METRIC_DOC =
+      "The total number of responses sent in the given window.";
+
+  // errors
+  static final String RECORD_ERROR_COUNT_WINDOWED_METRIC_NAME = "error-count-windowed";
+  private static final String RECORD_ERROR_COUNT_WINDOWED_METRIC_DOC =
+      "The total number of record sends that resulted in errors in the given window.";
+
+  protected static final String RECORD_ERROR_RATE_METRIC_NAME = "record-error-rate";
+  private static final String RECORD_ERROR_RATE_METRIC_DOC =
+      "The average per-second number of record sends that resulted in errors.";
+
+  // latency
+  static final String REQUEST_LATENCY_AVG_METRIC_NAME = "request-latency-avg";
+  private static final String REQUEST_LATENCY_AVG_METRIC_DOC = "The average request latency";
+
+  static final String REQUEST_LATENCY_MAX_METRIC_NAME = "request-latency-max";
+  private static final String REQUEST_LATENCY_MAX_METRIC_DOC = "The max request latency";
+
+  static final String REQUEST_LATENCY_PCT_METRIC_PREFIX = "request-latency-";
+  private static final String REQUEST_LATENCY_PCT_METRIC_DOC = "Request latency percentiles.";
+
+  protected Metrics metrics;
+  protected String jmxPrefix;
+  protected String requestSensorName;
+  protected String requestSizeSensorName;
+  protected String responseSensorName;
+  protected String recordErrorSensorName;
+  protected String requestLatencySensorName;
```

Comment:
One way to keep everything here `final` and to generally make the actually differing implementation as close to immutable as possible (if not entirely immutable) is to extract the polymorphic setup in `ProducerMetricsBuilder`, because this is more or less what we are ending up with.